File: backend/src/main/java/ma/investpro/service/FournisseurService.java
package ma.investpro.service;

import lombok.RequiredArgsConstructor;
import ma.investpro.dto.FournisseurDTO;
import ma.investpro.entity.Fournisseur;
import ma.investpro.exception.ResourceNotFoundException;
import ma.investpro.repository.FournisseurRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class FournisseurService {

    private final FournisseurRepository fournisseurRepository;

    public List<FournisseurDTO> getAll() {
        return fournisseurRepository.findAll().stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    public List<FournisseurDTO> getAllActive() {
        return fournisseurRepository.findByActifTrue().stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    public FournisseurDTO getById(Long id) {
        Fournisseur fournisseur = fournisseurRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Fournisseur", "id", id));
        return toDTO(fournisseur);
    }

    public List<FournisseurDTO> search(String search) {
        return fournisseurRepository.search(search).stream()
            .map(this::toDTO)
            .collect(Collectors.toList());
    }

    @Transactional
    public FournisseurDTO create(FournisseurDTO dto) {
        if (fournisseurRepository.existsByCode(dto.getCode())) {
            throw new IllegalArgumentException("Un fournisseur avec le code " + dto.getCode() + " existe déjà");
        }
        if (dto.getIce() != null && fournisseurRepository.existsByIce(dto.getIce())) {
            throw new IllegalArgumentException("Un fournisseur avec l'ICE " + dto.getIce() + " existe déjà");
        }

        Fournisseur fournisseur = toEntity(dto);
        fournisseur.setActif(true);
        Fournisseur saved = fournisseurRepository.save(fournisseur);
        return toDTO(saved);
    }

    @Transactional
    public FournisseurDTO update(Long id, FournisseurDTO dto) {
        Fournisseur fournisseur = fournisseurRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Fournisseur", "id", id));

        if (!fournisseur.getCode().equals(dto.getCode()) && fournisseurRepository.existsByCode(dto.getCode())) {
            throw new IllegalArgumentException("Un fournisseur avec le code " + dto.getCode() + " existe déjà");
        }
        if (dto.getIce() != null && !dto.getIce().equals(fournisseur.getIce()) && fournisseurRepository.existsByIce(dto.getIce())) {
            throw new IllegalArgumentException("Un fournisseur avec l'ICE " + dto.getIce() + " existe déjà");
        }

        updateEntityFromDTO(fournisseur, dto);
        Fournisseur updated = fournisseurRepository.save(fournisseur);
        return toDTO(updated);
    }

    @Transactional
    public void delete(Long id) {
        Fournisseur fournisseur = fournisseurRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Fournisseur", "id", id));
        fournisseur.setActif(false);
        fournisseurRepository.save(fournisseur);
    }

    private FournisseurDTO toDTO(Fournisseur entity) {
        return FournisseurDTO.builder()
            .id(entity.getId())
            .code(entity.getCode())
            .raisonSociale(entity.getRaisonSociale())
            .identifiantFiscal(entity.getIdentifiantFiscal())
            .ice(entity.getIce())
            .adresse(entity.getAdresse())
            .ville(entity.getVille())
            .telephone(entity.getTelephone())
            .fax(entity.getFax())
            .email(entity.getEmail())
            .contact(entity.getContact())
            .nonResident(entity.getNonResident())
            .remarques(entity.getRemarques())
            .actif(entity.getActif())
            .build();
    }

    private Fournisseur toEntity(FournisseurDTO dto) {
        return Fournisseur.builder()
            .code(dto.getCode())
            .raisonSociale(dto.getRaisonSociale())
            .identifiantFiscal(dto.getIdentifiantFiscal())
            .ice(dto.getIce())
            .adresse(dto.getAdresse())
            .ville(dto.getVille())
            .telephone(dto.getTelephone())
            .fax(dto.getFax())
            .email(dto.getEmail())
            .contact(dto.getContact())
            .nonResident(dto.getNonResident() != null ? dto.getNonResident() : false)
            .remarques(dto.getRemarques())
            .build();
    }

    private void updateEntityFromDTO(Fournisseur entity, FournisseurDTO dto) {
        entity.setCode(dto.getCode());
        entity.setRaisonSociale(dto.getRaisonSociale());
        entity.setIdentifiantFiscal(dto.getIdentifiantFiscal());
        entity.setIce(dto.getIce());
        entity.setAdresse(dto.getAdresse());
        entity.setVille(dto.getVille());
        entity.setTelephone(dto.getTelephone());
        entity.setFax(dto.getFax());
        entity.setEmail(dto.getEmail());
        entity.setContact(dto.getContact());
        entity.setNonResident(dto.getNonResident() != null ? dto.getNonResident() : false);
        entity.setRemarques(dto.getRemarques());
        if (dto.getActif() != null) {
            entity.setActif(dto.getActif());
        }
    }
}

File: backend/src/main/kotlin/ma/investpro/service/FournisseurService.kt
package ma.investpro.service

import ma.investpro.entity.Fournisseur
import ma.investpro.repository.FournisseurRepository
import mu.KotlinLogging
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

private val logger = KotlinLogging.logger {}

/**
 * Service for Fournisseur entity
 * Provides CRUD operations for suppliers
 */
@Service
@Transactional(readOnly = true)
class FournisseurService(
    private val fournisseurRepository: FournisseurRepository
) {

    @Transactional(readOnly = true)
    fun findAll(pageable: org.springframework.data.domain.Pageable): org.springframework.data.domain.Page<Fournisseur> {
        return fournisseurRepository.findAll(pageable)
    }

    @Cacheable("fournisseurs")
    fun findAll(): List<Fournisseur> {
        logger.debug { "Fetching all fournisseurs" }
        return fournisseurRepository.findAll().also { fournisseurs ->
            logger.info { "Found ${fournisseurs.size} fournisseurs" }
        }
    }

    fun findById(id: Long): Fournisseur {
        logger.debug { "Fetching fournisseur by ID: $id" }
        return fournisseurRepository.findById(id)
            .orElseThrow {
                logger.warn { "Fournisseur not found - ID: $id" }
                IllegalArgumentException("Fournisseur avec ID $id non trouvé")
            }
    }

    @Cacheable("fournisseurs-actifs")
    fun findAllActive(): List<Fournisseur> {
        logger.debug { "Fetching all active fournisseurs" }
        return fournisseurRepository.findByActifTrue().also { fournisseurs ->
            logger.info { "Found ${fournisseurs.size} active fournisseurs" }
        }
    }

    fun search(query: String): List<Fournisseur> {
        logger.debug { "Searching fournisseurs with query: $query" }
        return fournisseurRepository.findByRaisonSocialeContainingIgnoreCase(query).also { results ->
            logger.info { "Found ${results.size} fournisseurs matching '$query'" }
        }
    }

    @Transactional
    @CacheEvict(value = ["fournisseurs", "fournisseurs-actifs"], allEntries = true)
    fun save(fournisseur: Fournisseur): Fournisseur {
        logger.debug { "Saving fournisseur: ${fournisseur.code}" }

        // Validate unique code for new entities
        if (fournisseur.id == null) {
            if (fournisseurRepository.existsByCode(fournisseur.code)) {
                throw IllegalArgumentException("Un fournisseur avec le code '${fournisseur.code}' existe déjà")
            }
            if (!fournisseur.ice.isNullOrBlank() && fournisseurRepository.existsByIce(fournisseur.ice!!)) {
                throw IllegalArgumentException("Un fournisseur avec l'ICE '${fournisseur.ice}' existe déjà")
            }
        }

        return fournisseurRepository.save(fournisseur).also {
            logger.info { "Saved fournisseur with ID: ${it.id}" }
        }
    }
}
